src/Engine/Events/EventMap.cpp

```diff
@@ -1,12 +1,58 @@
 #include <utility>
 #include <vector>
 
+#include "Engine/Party.h"
 #include "Engine/Events/EventMap.h"
 #include "Engine/Engine.h"
+#include "Engine/Graphics/Level/Decoration.h"
+#include "Engine/mm7_data.h"
 #include "Utility/MapAccess.h"
 
-void EventMap::add(int id, EventIR ir) {
-    _eventsById[id].push_back(ir);
+void EventMap::add(int eventId, EventIR ir) {
+    _eventsById[eventId].push_back(ir);
+}
+
+void EventMap::execute(int eventId, int startStep, bool canShowMessages) const {
+    assert(_eventsById.contains(eventId));
+    assert(startStep >= 0);
+
+    int step = startStep;
+    bool stepFound;
+    PLAYER_CHOOSE_POLICY who = !pParty->hasActiveCharacter() ? CHOOSE_RANDOM : CHOOSE_ACTIVE;
+    bool mapExitTriggered;
+
+    do {
+        stepFound = false;
+        for (const EventIR &ir : valueOr(_eventsById, eventId, std::vector<EventIR>())) {
+            if (ir.step == step) {
+                step = ir.execute(canShowMessages, &who, &mapExitTriggered);
```

Comment:
Is `who` essentially a shared state between different events? One can set it, then the next one can use it?

Reply:
Got it, thanks!